spec.html

```diff
@@ -32,8 +32,118 @@ <h1>Well-known intrinsic objects</h1>
+	<emu-alg>
+		1. Let _target_ be _F_.[[WrappedTargetFunction]].
+		1. Assert: IsCallable(_target_) is *true*.
+		1. Let _targetRealm_ be ? GetFunctionRealm(_target_).
+		1. Let _callerRealm_ be ? GetFunctionRealm(_F_).
+		1. For each element _key_ of _argumentsList_, do
+			1. Let o be _argumentsList_[_key_].
+			1. Set _argumentsList_[_key_] to ? GetWrappedValue(_targetRealm_, _o_).
+		1. Let _wrappedThisArgument_ to ? GetWrappedValue(_targetRealm_, _thisArgument_).
+		1. Let _value_ be ? Call(_target_, _wrappedThisArgument_, _argumentsList_).
+		1. Return ? GetWrappedValue(_callerRealm_, _value_).
+	</emu-alg>
+	</emu-clause>
+
+	<emu-clause id="sec-wrappedfunctioncreate" aoid="WrappedFunctionCreate">
+	<h1>WrappedFunctionCreate ( _callerRealm_, _targetFunction_ )</h1>
+	<p>The abstract operation WrappedFunctionCreate takes arguments _callerRealm_ and _targetFunction_. It is used to specify the creation of new wrapped function exotic objects. It performs the following steps when called:</p>
+	<emu-alg>
+		1. Assert: _callerRealm_ is a Realm Record.
+		1. Assert: IsCallable(_targetFunction_) is *true*.
+		1. Let _internalSlotsList_ be the internal slots listed in <emu-xref href="#table-internal-slots-of-wrapped-function-exotic-objects"></emu-xref>, plus [[Prototype]] and [[Extensible]].
+		1. Let _obj_ be ! MakeBasicObject(_internalSlotsList_).
+		1. Set _obj_.[[Prototype]] to %Function.prototype% from _callerRealm_.
+		1. Set _obj_.[[Call]] as described in <emu-xref href="#sec-wrapped-function-exotic-objects-call-thisargument-argumentslist"></emu-xref>.
+		1. Set _obj_.[[Realm]] to _callerRealm_.
+		1. Set _obj_.[[WrappedTargetFunction]] to _targetFunction_.
+		1. Return _obj_.
+	</emu-alg>
+	</emu-clause>
+</emu-clause>
+
 <emu-clause id="sec-realm-objects">
 	<h1>Realm Objects</h1>
+	<emu-clause id="sec-realm-abstracts">
+		<h1>Realm Abstract Operations</h1>
+
+		<emu-clause id="sec-performrealmevaluation" aoid="PerformRealmEvaluation">
+			<h1>PerformRealmEvaluation ( _x_, _evalRealm_ )</h1>
+			<emu-alg>
+				1. Assert: Type(_x_) is String.
+				1. Assert: _evalRealm_ is a Realm Record.
+				1. Let _hostDefined_ be _evalRealm_.[[HostDefined]].
+				1. Let _s_ be ParseScript(_x_, _evalRealm_, _hostDefined_).
+				1. If _s_ is a List of errors, then
+					1. Perform HostReportErrors(_s_).
+					1. Return NormalCompletion(*undefined*).
+				1. Return ? ScriptEvaluation(_s_).
+			</emu-alg>
+		</emu-clause>
+
+		<emu-clause id="sec-getwrappedvalue" aoid="GetWrappedValue">
+			<h1>GetWrappedValue ( _callerRealm_, _value_ )</h1>
+			<emu-alg>
+				1. Assert: _callerRealm_ is a Realm Record.
+				1. If Type(_value_) is Object, then
+				  1. If IsCallable(_value_) is *false*, throw a TypeError exception.
+				  1. Return ? WrappedFunctionCreate(_callerRealm_, _value_).
```

Comment:
It will return a different function even the `_value_`-`_callerRealm_` pair is the same.

Reply:
that is by design as explained in https://github.com/tc39/proposal-realms/issues/289, this is not a membrane, it does not do identity reconciliation. Every time it sees a callable coming thru the boundary, a new wrapped function is created.

Reply:
If we don't keep identity reconciliation, we will never be able to make it after we ship it cause it's breaking change. That will block the possibility to the membrane

Reply:
No, that's not true, you can implement identity preserving membrane on top of this.

Reply:
It would be the responsibility of the membrane or library sitting on top of this API to avoid recreating wrappers.

Given that symbols are unique and comparable through round trips, wrapper functions provide indirect cross realms references, and WeakRefs are available, it should be possible to build a full non-leaky proxy based membrane.  I believe the next step is to validate this by building such a membrane on top of this proposed API. See https://github.com/leobalter/realms-polyfill/issues/13

Reply:
One question however is if the `GetWrappedValue` operation should automatically unwrap a wrapped function if its `[[Realm]]` matches `callerRealm`. Right now it would double wrap it.

Reply:
I'm opening an issue to discuss if we should unwrap or double wrap it.

I don't see much value or the goal in sending the same function back and forth. If the goal is to get a unique identity, symbols should resolve this trick in a much more lightweight manner. 

Adding an extra check for every function to see if it can be unwrapped might end with some cost, but at the same time, it cuts a chain of references here and there.